pkg/kubelet/lifecycle/interfaces.go

```diff
@@ -0,0 +1,70 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package lifecycle
+
+import "k8s.io/kubernetes/pkg/api"
+
+// PodAdmitAttributes is the context for a pod admission decision.
+// The member fields of this struct should never be mutated.
+type PodAdmitAttributes struct {
+	// the pod to evaluate for admission
+	Pod *api.Pod
+	// all pods bound to the kubelet, includes the pod being evaluated.
+	Pods []*api.Pod
+}
+
+// PodAdmitResult provides the result of a pod admission decision.
+type PodAdmitResult struct {
+	// if true, the pod should be admitted.
+	Admit bool
+	// a brief single-word reason why the pod could not be admitted.
+	Reason string
+	// a brief message explaining why the pod could not be admitted.
+	Message string
+}
+
+// PodAdmitter is an interface invoked by the kubelet during admission
+// decisions.
+type PodAdmitter interface {
+	// Admit evaluates if the incoming pod can be admitted to the kubelet.
+	Admit(attrs *PodAdmitAttributes) PodAdmitResult
+}
+
+// PodEvictAttributes is the context for a pod eviction decision.
+// The member fields of this struct should never be mutated.
+type PodEvictAttributes struct {
+	// the pod to evaluate for eviction.
+	Pod *api.Pod
+}
+
+// PodEvictResult provides the result of a pod eviction decision.
+type PodEvictResult struct {
+	// if true, the pod should be evicted.
+	Evict bool
+	// a brief single-word reason why the pod should be evicted.
+	Reason string
+	// a brief single-word reason why the pod should be evicted.
+	Message string
+}
+
+// PodEvictor is an interface invoked by the kubelet to evaluate if a pod
```

Comment:
I think I am biased to keep this on a per pod call out given the current code flow since each integration point today is already nested in an iteration loop over the pod set.  Since this is internal to the kubelet today, if we find a later refactor justifies a list returning a list pattern I think we can update at that time.